pkg/objstore/cos/cos.go

```diff
@@ -0,0 +1,366 @@
+}
+
+// Name returns the bucket name for s3.
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	if _, err := b.client.Object.Put(ctx, name, r, nil); err != nil {
+		return errors.Wrap(err, "upload cos object")
+	}
+	return nil
+}
+
+// Delete removes the object with the given name.
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	if _, err := b.client.Object.Delete(ctx, name); err != nil {
+		return errors.Wrap(err, "delete cos object")
+	}
+	return nil
+}
+
+// Iter calls f for each entry in the given directory (not recursive.). The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, dirDelim) + dirDelim
+	}
+
+	for object := range b.ListObjects(ctx, dir, false) {
+		if object.err != nil {
+			return object.err
+		}
+		if object.key == "" {
+			continue
+		}
+		if err := f(object.key); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (b *Bucket) getRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	if len(name) == 0 {
+		return nil, errors.Errorf("given object name should not empty")
+	}
+
+	opts := &cos.ObjectGetOptions{}
+	if length != -1 {
+		if err := setRange(opts, off, off+length-1); err != nil {
+			return nil, err
+		}
+	}
+
+	resp, err := b.client.Object.Get(ctx, name, opts)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := resp.Body.Read(nil); err != nil {
+		runutil.CloseWithLogOnErr(b.logger, resp.Body, "cos get range obj close")
+		return nil, err
+	}
+
+	return resp.Body, nil
+}
+
+// Get returns a reader for the given object name.
+func (b *Bucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return b.getRange(ctx, name, 0, -1)
+}
+
+// GetRange returns a new range reader for the given object name and range.
+func (b *Bucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	return b.getRange(ctx, name, off, length)
+}
+
+// Exists checks if the given object exists in the bucket.
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	if _, err := b.client.Object.Head(ctx, name, nil); err != nil {
+		if b.IsObjNotFoundErr(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "head cos object")
+	}
+
+	return true, nil
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	tmpErr := toErrorResponse(err)
+	if tmpErr.Code == "NoSuchKey" {
+		return true
+	}
+	if tmpErr.Response != nil && tmpErr.Response.StatusCode == http.StatusNotFound {
+		return true
+	}
+	return false
+}
+
+func toErrorResponse(err error) *cos.ErrorResponse {
```

Comment:
this is kind of misleading function. In scope of `IsObjNotFoundErr` it makes sense, but for other potential usages, you could be misleaded that in case of this error is not cos.ErrorResposne you are getting empty error struct. Returning nil would be better. But since you use it only in `IsObjNotFoundErr` I would just copy this logic into `IsObjNotFoundErr` and return false 

What do you think?

Reply:
Agree, thanks!